### PR TITLE
[jest-gas-reporter] Store gas consumption as test artifacts

### PR DESCRIPTION
**Pull request by geoknee**
This PR modifies the reporter to store gas consumption information as an object of type`GasCosts`. This is 
1. written to `gas.json` 
2. formatted as an `easy-table` and
    - `console.logged`
    - formatted as a markdown table and written to `gas.md` 

Currently `gas.json` and `gas.md` are i) always appended to when tests run; ii) `gitignored` and iii) uploaded as artifacts on circle. The ideas is that `gas.md` is easy on the eye, and `gas.json` will make it easy to read in the data if we want to trawl through the artifacts and track the gas consumption as we iterate on the contracts or modify the evm version.

### CODE REVIEW

File: packages/jest-gas-reporter/src/index.ts [hunk 1]
```diff
@@ -91,6 +98,9 @@ export class GasReporter implements jest.Reporter {
       this.options.contractArtifactFolder
     );
     this.outputGasInfo(contractCalls);
+    // if (process.env.CIRCLECI) {
+    await this.saveResultsToFile(process.env.CIRCLE_SHA1);
+    // }
```
Comment (geoknee) [inline packages/jest-gas-reporter/src/index.ts]: To be removed or reinstated depending on PR comments.

File: packages/jest-gas-reporter/src/index.ts [hunk 2]
```diff
@@ -253,6 +272,35 @@ export class GasReporter implements jest.Reporter {
...
+      revision: hash,
+      gasConsumed
+    };
+    const resultsString = JSON.stringify(results, null, 4) + '\n';
+    await fs.appendFile('./gas.json', resultsString, err => {
+      if (err) throw err;
+      console.log('Wrote json to gas.json');
+    });
+    await fs.appendFile(
+      './gas.txt',
```
Comment (snario) [inline packages/jest-gas-reporter/src/index.ts]: Would it be possible for this file to be in Markdown format? Most places where `.txt` is readable, `.md` is also readable.

File: packages/rps/config/jest/jest.config.js [hunk 3]
```diff
@@ -41,4 +41,5 @@ module.exports = {
     'node',
     'mjs',
   ],
+  roots: ['<rootDir>/src'],
```
Comment (snario) [inline packages/rps/config/jest/jest.config.js]: Why is this necessary?

File: packages/jest-gas-reporter/src/index.ts [hunk 4]
```diff
@@ -164,49 +168,64 @@ export class GasReporter implements jest.Reporter {
...
+
+    console.log(this.objectToTable(gasConsumed).toString());
+  }
+
+  objectToTable(gasConsumed: GasConsumed): easyTable {
+    const table = new easyTable();
+    Object.keys(gasConsumed).forEach(contract => {
+      const contractStats = gasConsumed[contract];
+      table.cell('Contract', contract);
+      table.cell('Deployment', contractStats.deployment);
```
Comment (snario) [inline packages/jest-gas-reporter/src/index.ts]: Why is this value relevant? Also what is it exactly?

File: packages/jest-gas-reporter/src/index.ts [hunk 5]
```diff
@@ -253,6 +272,35 @@ export class GasReporter implements jest.Reporter {
       }
     }
   }
+
+  async saveResultsToFile(hash: string): Promise<void> {
+    const results = {
+      date: Date.now(),
```
Comment (snario) [inline packages/jest-gas-reporter/src/index.ts]: You could format this with `Date(Date.now()).toString()`

File: packages/jest-gas-reporter/src/index.ts [hunk 6]
```diff
@@ -253,6 +272,35 @@ export class GasReporter implements jest.Reporter {
...
+    };
+    const resultsString = JSON.stringify(results, null, 4) + '\n';
+    await fs.appendFile('./gas.json', resultsString, err => {
+      if (err) throw err;
+      console.log('Wrote json to gas.json');
+    });
+    await fs.appendFile(
+      './gas.txt',
+      '\n\n\ndate: ' +
+        Date.now() +
```
Comment (snario) [inline packages/jest-gas-reporter/src/index.ts]: You could format this with `Date(Date.now()).toString()`
```suggestion
        Date(Date.now()).toString() +
```

File: packages/rps/config/jest/jest.config.js (hunk 3 above)
Comment (geoknee) [reply packages/rps/config/jest/jest.config.js]: It prevents jest from finding the compiled version of any mocks (source in `/src/`, output in `build`). Without this, it finds both and complains about a duplicate.

File: packages/jest-gas-reporter/src/index.ts (hunk 4 above)
Comment (geoknee) [reply packages/jest-gas-reporter/src/index.ts]: Gas consumed when deploying the contract.

Comment (snario) [reply packages/jest-gas-reporter/src/index.ts]: Ohhh okay.

File: .circleci/config.yml [hunk 7]
```diff
@@ -104,6 +105,14 @@ jobs:
       - upload_logs:
           file: test-stats
       - notify_slack
+      - store_artifacts:
```
Comment (andrewgordstewart) [inline .circleci/config.yml]: You could extract this into a command, like:
```
  upload_gas_report:
    description: 'Upload gas report '
    parameters:
      package:
        type: string
    steps:
      - store_artifacts:
          path: /home/circleci/packages/<< parameters.package >>./gas.json
          destination: << parameters.package >>-gas.json
      - store_artifacts:
          path: /home/circleci/packages/<< parameters.package >>./gas.txt
          destination: << parameters.package >>-gas.txt
```

then, this could be:
```
- upload_gas_report:
  package: rps
- upload_gas_report:
  package: nitro-protocol
```
making it a little easier to maintain
